### PR TITLE
Add composition support to `prove_bisim`

### PR DESCRIPTION
**Pull request by bboston7**
This change enables the reuse of proofs in `prove_bisim`.  `prove_bisim` now returns a `BisimTheorem` that can be passed in to future `prove_bisim` commands to facilitate compositional proofs.  To support this, `prove_bisim` now also requires an additional relation over states.  For a detailed description of how this all works, see the module level doc comment in `SAWScript.Bisimulation`.

### CODE REVIEW

File: src/SAWScript/Bisimulation.hs [hunk 1]
```diff
@@ -212,7 +664,77 @@ proveBisimulation script relation lhs rhs = do
           , "  Expected: (" ++ stStr ++ ", inputType) -> (" ++ stStr ++ ", outputType)"
           , "  Actual: " ++ show (ppTypedTermType (ttType side)) ]
 
-    -- Generate 'Term' for application of a relation
-    scRelation :: SharedContext -> Term -> Term -> TopLevel Term
-    scRelation sc relLhs relRhs = io $
-      scApplyAll sc (ttTerm relation) [relLhs, relRhs]
+-- | Replace the invocation of a specific 'Constant' with an 'ExtCns'.  The function returns the resulting 'Term' and updates a 'ReplaceState' to hold the generated 'ExtCns' and the specific 'App' that was replaced.
```
Comment (RyanGlScott) [inline src/SAWScript/Bisimulation.hs:667]: Wrap this around to ~80-column lines.

File: src/SAWScript/Bisimulation.hs [hunk 2]
```diff
@@ -13,144 +13,561 @@ functionality in the future.
...
+the state relation of the outer terms must ensure that the inner terms' state
+relation holds.  As such, the prover generates and proves a side condition for
+each applied 'BisimTheorem'.  Let @g_lhs_s@ and @g_rhs_s@ be the states for
+@g_lhs@ and @g_rhs@ respectively.  Additionally, let @f_lhs_s@ be the state for
+@f_lhs@ expressed as a substate of @g_rhs_s@, and @f_rhs_s@ be the state for
+@f_rhs@ expressed as a substate of @g_rhs_s@ (the prover examines the call sites
+for @f_lhs@ and @f_rhs@ within @g_lhs@ and @g_rhs@ to extract these @f@ states
+in terms of the @g@ states).  Lastly, let @g_srel@ be the state relation for the
+@g@ functions, and @f_srel@ be the state relation for the @f@ functions.  The
+prover then checks:
```
Comment (RyanGlScott) [inline src/SAWScript/Bisimulation.hs:56]: This is a pretty long paragraph. Is there a way to split it up more?

File: src/SAWScript/Bisimulation.hs [hunk 3]
```diff
@@ -13,144 +13,561 @@ functionality in the future.
...
+ -- ^ Right hand side of the bisimulation
+  , bcOutputType :: C.Type
+ -- ^ Output type of the bisimilar functions
+  , bcLhsStateType :: C.Type
+ -- ^ State type for 'bcLeftSide'
+  , bcRhsStateType :: C.Type
+ -- ^ State type for 'bcRightSide'
+  , bcInputType :: C.Type
+ -- ^ Input type of the bisimilar functions
+  }
```
[the remark this answers is not in the document]
Comment (bboston7) [reply src/SAWScript/Bisimulation.hs]: Good idea. I changed `BisimComponents` to hold a `BisimTheorem`.

File: doc/manual/manual.md [hunk 4]
```diff
@@ -4217,41 +4235,93 @@ nandSpec (_, (x, y)) = ((), (True, ~ (andSpec ((), (x, y))).1.1))
...
+terms simulate eachother.
+
+When using composition, the `prove_bisim` also generates and attempts to solve
+the proof goal below for any successfully applied `BisimTheorem` in the
+`theorems` list:
+
+~~~~
+COMPOSITION SIDE CONDITION:
+  forall g_lhs_s g_rhs_s.
+    g_srel g_lhs_s g_rhs_s -> f_srel f_lhs_s f_rhs_s
```
[the remark this answers is not in the document]
Comment (bboston7) [reply doc/manual/manual.md:4314]: This was confusing. `f_lhs_s` and `f_rhs_s` are substates of `g_lhs_s` and `g_rhs_s` respectively, so they should not be quantified over.  I added a helper function `extract_substate` to the documentation and now use that to set `f_lhs_s` and `f_rhs_s` in terms of the other variables.  That hopefully makes it more clear what's going on here.

File: src/SAWScript/Bisimulation.hs (hunk 2 above)
Comment (bboston7) [reply src/SAWScript/Bisimulation.hs]: I broke this up a bit when I added some clarification around `f_lhs_s` and `f_rhs_s`

File: doc/manual/manual.md [hunk 5]
```diff
@@ -4217,41 +4235,93 @@ nandSpec (_, (x, y)) = ((), (True, ~ (andSpec ((), (x, y))).1.1))
...
+~~~~
+
+where `g_lhs` contains a call to `f_lhs` and `g_rhs` contains a call to `f_rhs`, and the variables in `COMPOSITION SIDE CONDITION` are:
```
[the remark this answers is not in the document]
Comment (bboston7) [reply doc/manual/manual.md]: That's correct. I added some more documentation to clarify that.

File: doc/manual/manual.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bboston7) [reply doc/manual/manual.md:4314]: I wound up renaming `extract_substate` to `extract_inner_state` for consistency

**Review by RyanGlScott — APPROVED** [review]
I like the overall look of this PR, with the caveat that I have not looked deeply into the implementation of the new feature. Let me know if there's something that deserves a closer look.

**Pull request merged**